### PR TITLE
Cleanup process emulator

### PR DESCRIPTION
**Pull request by lightsofapollo**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -15,7 +15,7 @@
     "prettier": "^1.12.1"
   },
   "scripts": {
-    "test": "yarn jest --coverage",
+    "test": "yarn jest --coverage --detectOpenHandles",
```
Comment (cbaron) [inline package.json]: Just discovered this too.

File: packages/appsync-emulator-serverless/serverCore.js [hunk 2]
```diff
@@ -31,18 +31,28 @@ class SubscriptionServer {
     mqttServer.on('clientDisconnected', (...args) =>
       this.onClientDisconnect(...args),
     );
+
+    // XXX: Note we may need to listen to unsubscribe to end the async iterator.
```
Comment (cbaron) [inline packages/appsync-emulator-serverless/serverCore.js]: Agreed.

In the request for price tool I can foresee a customer potentially unsubscribing when the current transaction is completed.

File: packages/appsync-emulator-serverless/serverCore.js [hunk 3]
```diff
@@ -31,18 +31,28 @@ class SubscriptionServer {
...
   async onClientConnect(client) {
     const { id: clientId } = client;
-    log.info('clientConnect', { clientId });
     consola.info(`client connected to subscription server (${clientId})`);
+  }
+
+  async onClientSubscribed(topic, client) {
+    const { id: clientId } = client;
+    consola.info(`client (${clientId}) subscribed to : ${topic}`);
+    log.info(`client (${clientId}) subscribed to : ${topic}`);
```
Comment (cbaron) [inline packages/appsync-emulator-serverless/serverCore.js]: why `consola` and `log` ?

File: packages/appsync-emulator-serverless/tester.js [hunk 4]
```diff
@@ -41,10 +28,10 @@ const create = async ({ serverless, schemaPath, port = 0 } = {}) => {
     subscriptions,
   });
 
-  const close = () => {
+  const close = async () => {
     server.close();
     // schema deletes tables so we must close the emulator after.
-    return schemaClose();
+    await schemaClose();
```
Comment (cbaron) [inline packages/appsync-emulator-serverless/tester.js]: Is this a stylistic change, or something to make jest happy?

File: packages/dynamodb-emulator/client.js [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+      'ignore',
+      fs.openSync(path.join(pidPath, `${hash}-stdout.log`), 'w'),
+      fs.openSync(path.join(pidPath, `${hash}-stderr.log`), 'w'),
+    ],
+  });
+  proc.unref();
+
+  const start = Date.now();
+  while (Date.now() - start < waitTimeout) {
+    if (await fsExists(unixSocketFile)) {
```
Comment (cbaron) [inline packages/dynamodb-emulator/client.js]: Curious why you didn't use `fs.existsSync`

File: packages/dynamodb-emulator/deamon.js [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+      log.error(err);
+      server.close();
+    }
+  }
+  resolvedGlobalEmulator = emu;
+  emu.proc.once('exit', () => {
+    server.close();
+  });
+  return emu;
+}
```
Comment (cbaron) [inline packages/dynamodb-emulator/deamon.js]: Shouldn't `globalEmulator` get assigned to resolvedGlobalEmulator?

File: packages/dynamodb-emulator/deamon.js [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+function cleanup() {
+  server.close();
+  try {
+    fs.unlinkSync(pidFile);
+    fs.unlinkSync(unixSocket);
+    // eslint-disable-next-line
+  } catch (err) {
+    // we don't care if there was an error in this case.
+    // keep cleaning up!
```
Comment (cbaron) [inline packages/dynamodb-emulator/deamon.js]: Might be useful to add a log here ?

File: packages/appsync-emulator-serverless/serverCore.js (hunk 3 above)
Comment (lightsofapollo) [reply packages/appsync-emulator-serverless/serverCore.js]: One logs to stdout without specifying NODE_DEBUG and the other logs to stderr and only with NODE_DEBUG.

File: packages/appsync-emulator-serverless/tester.js (hunk 4 above)
Comment (lightsofapollo) [reply packages/appsync-emulator-serverless/tester.js]: stylistic (result of some reverted refactoring)

File: packages/dynamodb-emulator/client.js (hunk 5 above)
Comment (lightsofapollo) [reply packages/dynamodb-emulator/client.js]: I didn't want to use a synchronous method in a while true loop . Easy to effectively deadlock node this way.

File: packages/dynamodb-emulator/deamon.js (hunk 6 above)
Comment (lightsofapollo) [reply packages/dynamodb-emulator/deamon.js]: I intentionally kept the data types consistent so globalEmulator is always a promise (even if it's resolved) resolvedGlobalEmulator is sort of hacky but is there for graceful cleanup.

File: packages/dynamodb-emulator/deamon.js (hunk 7 above)
Comment (lightsofapollo) [reply packages/dynamodb-emulator/deamon.js]: +1

File: packages/appsync-emulator-serverless/serverCore.js (hunk 2 above)
Comment (lightsofapollo) [reply packages/appsync-emulator-serverless/serverCore.js]: I am going to double check but I think this current implementation is fine ... We essentially end up with one connection per subscription . I will verify that works.

**Pull request merged**